client/src/pages/Challenges.tsx

```diff
@@ -0,0 +1,197 @@
+     * @returns the grid of ChallengeCards
+     */
+    const makeGrid = useCallback((): JSX.Element[] => {
+        if (challengesData.current === undefined) {
+            return [];
+        }
+
+        const grid: JSX.Element[] = [];
+        let row: JSX.Element[] = [];
+        let i = 0;
+        for (const challenge of challengesData.current) {
+            //make sure the challenge is not filtered out
+            if(!(filterIncluded(challenge.difficulty, filter) || filter === 0)) {
+                // console.log("Filtering out: " + challenge.difficulty);
+                continue;
+            }
+
+            row.push(
+                <Col lg={4} key={i} className="mb-4">
+                    <ChallengeCard {...challenge} />
+                </Col>
+            );
+            i++;
+            if (i % 3 === 0) {
+                grid.push(<Row sm={1} lg={3} key={i}>{row}</Row>);
+                row = [];
+            }
+        }
+        if (row.length > 0) {
+            grid.push(<Row key={i}>{row}</Row>);
+        }
+        return grid;
+    }, [filter]);
+
+    useEffect(() => {   
+        //fetch data about the challenge with the provided "id"
+        setIsLoading(true);
+        fetch('/api/challenges').then((response) => {
+            if(!response.ok){
+                throw new Error(response.statusText);                
+            }
+            return response.json() as Promise<ChallengeDetailsShort[]>; //THIS LINE MAY CAUSE ERRORS. Need to test with proper server.
+        }).then((data) => {
+            challengesData.current = data;
+            console.log(data);
+            setIsLoading(false);
+            return;
+        }).catch((err) => {
+            console.log();
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+            console.error("Failed fetching the challenges." + "\nError message: " + err.message)
+        });
+   }, [challengesData]);
+
+    useEffect(() => {    
+        if (challengesData.current != undefined) {
+            const sortedChallengesData = handleSortByDifficulty(challengesData.current);
+            challengesData.current = sortedChallengesData;
+        }
+        const newGrid = makeGrid();
+        setGrid(newGrid);
+        console.log("Grid updated");
+    }, [hideComplete, makeGrid, handleSortByDifficulty, challengesData, isLoading]);
+
+    
+    /**
+     * Retuns true if the difficulty is included in the filter
+     */
+    function filterIncluded(difficulty: ChallengeDifficulties, filter: number): boolean {
+        console.log("Filter: " + filter);
+        switch(difficulty) {
+            case ChallengeDifficulties.EASY:
+                return (filter & 1) === 1;
+            case ChallengeDifficulties.MEDIUM:
+                return (filter & 2) === 2;
+            case ChallengeDifficulties.HARD:
+                return (filter & 4) === 4;
+            default:
+                return false;
+        }
+    }
```

Comment:
How does this function work?

Reply:
It uses js bitwise operators: https://www.w3schools.com/js/js_bitwise.asp

You have a binary string of length 3.
`xxx`
Each bit correspond to a filter. First bit is easy, second bit is medium, third is hard. So this `101` would be easy and hard. Convert that to decimal you get `5`. The bitwise operator checks what bits the current filter has.

Reply:
Wouldn't it be simpler and more readable by tracking the filter with a list?
Example: If filter includes easy and hard
[ChallengeDifficulties.EASY, ChallengeDifficulties.HARD]

Then to check if difficulty is in the filter you can just return `array.includes(difficulty)`

----
After taking a look at how filter is implemented, wouldn't it be easier / more readable if we just had a `displayedDifficulties` state variable of type `ChallengeDifficulties[]` and then update that?

I say this as lines such as `filterIncluded(ChallengeDifficulties.EASY, filter) ? setFilter(filter & 6) : setFilter(filter | 1);` are difficult for readers to understand and mantain.

Now for toggling we create a function that updates the state
Pseudo code
```
if difficulty in filter:
    remove difficulty from array
else:
    add difficulty to array
update the filter state with setState
```

Reply:
That is what I thought initially, but I feel like the binary implementation is a lot more concise and, at least for me, easier to understand. I based it off of the linux permission commands, eg. `chmod 777` to give full access or `chmod 555` to give all users the read and execute permissions.

Reply:
For debugging purposes, you have a single value that you can print, without looping over a list. Plus the list would be printed containing `0, 1, 2`, as the difficulties are enums.

Reply:
We are not in that constrained of an environment. I would strongly suggest using strings as enums as opposed to individual bits.